src/persist-client/src/internal/gc.rs

```diff
@@ -205,273 +209,224 @@ where
         machine: &mut Machine<K, V, T, D>,
         req: GcReq,
     ) -> RoutineMaintenance {
-        // There's also a bulk delete API in s3 if the performance of this
-        // becomes an issue. Maybe make Blob::delete take a list of keys?
-        //
-        // https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html
-        async fn delete_all(
-            blob: &(dyn Blob + Send + Sync),
-            keys: impl Iterator<Item = BlobKey>,
-            metrics: &RetryMetrics,
-            span: Span,
-            semaphore: &Semaphore,
-        ) {
-            let futures = FuturesUnordered::new();
-            for key in keys {
-                futures.push(
-                    retry_external(metrics, move || {
-                        let key = key.clone();
-                        async move {
-                            let _permit = semaphore
-                                .acquire()
-                                .await
-                                .expect("acquiring permit from open semaphore");
-                            blob.delete(&key).await.map(|_| ())
-                        }
-                    })
-                    .instrument(span.clone()),
-                )
-            }
-
-            futures.collect().await
-        }
-
-        let delete_semaphore = Semaphore::new(
-            machine
-                .applier
-                .cfg
-                .dynamic
-                .gc_blob_delete_concurrency_limit(),
-        );
-
         let mut step_start = Instant::now();
         let mut report_step_timing = |counter: &Counter| {
             let now = Instant::now();
             counter.inc_by(now.duration_since(step_start).as_secs_f64());
             step_start = now;
         };
-
         assert_eq!(req.shard_id, machine.shard_id());
-        // NB: Because these requests can be processed concurrently (and in
-        // arbitrary order), all of the logic below has to work even if we've
-        // already gc'd and truncated past new_seqno_since.
 
         let mut states = machine
             .applier
             .state_versions
-            .fetch_all_live_states::<T>(req.shard_id)
+            .fetch_all_live_states(req.shard_id)
             .await
-            .expect("gc should only be called on an initialized shard")
-            // TODO: Consider pulling the K, V, D params off of GC. If we do,
-            // then we should be able to delete TypedStateVersionsIter (and
-            // probably merge UntypedStateVersionsIter into StateVersionsIter).
```

Comment:
i've forgotten all context here: should we do the second part of this todo (merge UntypedStateVersionsIter into StateVersionsIter), too?